server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -187,10 +215,19 @@ private void releaseDelayedOperations() {
                 queuedActions = new ArrayList<>(delayedOperations);
                 delayedOperations.clear();
             } else {
-                queuedActions = Collections.emptyList();
+                queuedActions = new ArrayList<>();
+                // Execute the next async block operation if there is one
+                final Optional<DelayedOperation> nextAsyncOperation = delayedOperations.stream()
+                    .filter(op -> op instanceof AsyncBlockDelayedOperation)
+                    .findFirst();
+                if (nextAsyncOperation.isPresent()) {
```

Comment:
I was happy with the simplicity of this change but there is a pitfall here: if a block op sneaks in and terminates it "releases" another async block op.